pkg/agent/nodeportlocal/portcache/port_table_windows.go

```diff
@@ -0,0 +1,298 @@
+		port := pt.PortSearchStart + i
+		if port > pt.EndPort {
+			// handle wrap around
+			port = port - numPorts
+		}
+		if _, ok := pt.NodePortTable[port]; ok {
+			// port is already taken
+			continue
+		}
+
+		protocols, err := addruleForPort(pt.PodPortRules, port, podIP, podPort, protocol)
+		if err != nil {
+			klog.V(4).InfoS("Port cannot be reserved, moving on to the next one", "port", port)
+			updateCloseState(protocols)
+			continue
+		}
+
+		pt.PortSearchStart = port + 1
+		if pt.PortSearchStart > pt.EndPort {
+			pt.PortSearchStart = pt.StartPort
+		}
+		return port, protocols, nil
+	}
+	return 0, nil, fmt.Errorf("no free port found")
+}
+
+//tryAddRuleforFreePort > getFreePort
+//This is the new windows function
+func (pt *PortTable) AddRule(podIP string, podPort int, protocol string) (int, error) {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	npData := pt.getEntryByPodIPPort(podIP, podPort)
+	exists := (npData != nil)
+	if !exists {
+		nodePort, protocols, err := pt.addRuleforFreePort(podIP, podPort, protocol)
+		//nodePort, protocols, err := pt.getFreePort(podIP, podPort)
+		//success means port, protocol available, state is open.
+		if err != nil {
+			return 0, err
+		}
+		npData = &NodePortData{
+			NodePort:  nodePort,
+			PodIP:     podIP,
+			PodPort:   podPort,
+			Protocols: protocols,
+		}
+		protocolSocketData := npData.FindProtocol(protocol)
+		protocolSocketData.State = stateInUse
+
+		pt.NodePortTable[nodePort] = npData
+		pt.PodEndpointTable[podIPPortFormat(podIP, podPort)] = npData
+	} else {
+		protocolSocketData := npData.FindProtocol(protocol)
+		if protocolSocketData == nil {
+			return 0, fmt.Errorf("unknown protocol %s", protocol)
+		}
+		if protocolSocketData.State == stateInUse {
+			return 0, fmt.Errorf("rule for %s:%d:%s already exists", podIP, podPort, protocol)
+		}
+		if protocolSocketData.State == stateClosed {
+			return 0, fmt.Errorf("invalid socket state for %s:%d:%s", podIP, podPort, protocol)
+		}
+
+		nodePort := npData.NodePort
+		if err := pt.PodPortRules.AddRule(nodePort, podIP, podPort, protocol); err != nil {
+			return 0, err
+		}
+
+		protocolSocketData.State = stateInUse
+	}
+	return npData.NodePort, nil
+}
+
+// syncRules ensures that contents of the port table matches the iptables rules present on the Node.
+func (pt *PortTable) syncRules() error {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	nplPorts := make([]rules.PodNodePort, 0, len(pt.NodePortTable))
+	for _, npData := range pt.NodePortTable {
+		protocols := make([]string, 0, len(supportedProtocols))
+		for _, protocol := range npData.Protocols {
+			if protocol.State == stateInUse {
+				protocols = append(protocols, protocol.Protocol)
+			}
+		}
+		nplPorts = append(nplPorts, rules.PodNodePort{
+			NodePort:  npData.NodePort,
+			PodPort:   npData.PodPort,
+			PodIP:     npData.PodIP,
+			Protocols: protocols,
+		})
+	}
+	if err := pt.PodPortRules.AddAllRules(nplPorts); err != nil {
```

Comment:
I think Windows netnatstaticmapping operation is not atomic, so it might return error with retry if one rule is added for 2nd time.

I would prefer to add rules one by one, and modify the state after success. And add state check in each rule before adding to ensure the atomic operations.

Reply:
The function will search the rule before insertion. So in this case I think it will find one existed rule and return nil.

Reply:
Updated. Removed unnecessary syncup function.